File: src/server/chain/gecko.ts
/**
 * GeckoTerminal's public API (no key). The free tier allows about 30 calls a
 * minute, but a shared cloud IP often gets less, so the pace adapts: it slows
 * down on every 429 and creeps back up while calls succeed.
 */
export interface GeckoPool {
  address: string; // as GeckoTerminal lists it: pair/pool address, or the v4 pool id
  dex: string;
  createdAt: string;
  baseToken: string;
  quoteToken: string;
  name: string;
  fdvUsd: number; // current fully diluted value of the base token
}

export interface GeckoToken {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
  supply: number; // normalized total supply
  imageUrl?: string;
}

export interface GeckoOptions {
  network: string;
  api?: string;
  perMinute?: number; // the ceiling; the actual pace adapts below it
  fetchImpl?: typeof fetch;
  sleep?: (ms: number) => Promise<void>;
  now?: () => number;
  log?: (msg: string) => void;
}

interface Resource<A> {
  attributes: A;
  relationships?: Record<string, { data?: { id?: string } }>;
}

const BATCH = 30; // the multi endpoints accept up to 30 addresses
const MIN_PER_MINUTE = 4;
const stripNetwork = (id = '') => id.slice(id.indexOf('_') + 1).toLowerCase();

export class GeckoClient {
  readonly stats = { calls: 0, throttled: 0, failed: 0, perMinute: 0 };
  private nextAt = 0;
  private readonly ceiling: number;
  private pace: number;

  constructor(private readonly o: GeckoOptions) {
    this.ceiling = o.perMinute ?? 28;
    this.pace = this.ceiling;
    this.stats.perMinute = this.pace;
  }

  private now() {
    return this.o.now?.() ?? Date.now();
  }

  private wait(ms: number) {
    return ms > 0 ? (this.o.sleep ?? ((t) => new Promise((resolve) => setTimeout(resolve, t))))(ms) : Promise.resolve();
  }

  /** How long a call made now would wait for its turn. */
  backlogMs() {
    return Math.max(0, this.nextAt - this.now());
  }

  private async get<T>(path: string): Promise<T | null> {
    for (let attempt = 0; attempt < 6; attempt++) {
      const at = Math.max(this.now(), this.nextAt);
      this.nextAt = at + 60_000 / this.pace;
      await this.wait(at - this.now());
      this.stats.calls++;
      try {
        const res = await (this.o.fetchImpl ?? fetch)(`${this.o.api ?? 'https://api.geckoterminal.com/api/v2'}/networks/${this.o.network}${path}`, {
          headers: { accept: 'application/json' },
        });
        if (res.status === 429) {
          this.stats.throttled++;
          this.pace = Math.max(MIN_PER_MINUTE, this.pace * 0.7);
          this.nextAt = Math.max(this.nextAt, this.now() + 10_000);
          this.stats.perMinute = Math.round(this.pace * 10) / 10;
          continue;
        }
        this.pace = Math.min(this.ceiling, this.pace + 0.2);
        this.stats.perMinute = Math.round(this.pace * 10) / 10;
        if (res.status === 404) return null;
        if (!res.ok) throw new Error(`HTTP ${res.status}`);
        return (await res.json()) as T;
      } catch (err) {
        this.stats.failed++;
        this.o.log?.(`geckoterminal ${path.slice(0, 60)}: ${(err as Error).message}`);
        this.nextAt = Math.max(this.nextAt, this.now() + 5_000 * (attempt + 1));
      }
    }
    // Unanswered is not the same as "no data": callers must retry later instead of drawing conclusions.
    throw new Error(`geckoterminal unavailable for ${path.slice(0, 60)}`);
  }

  /** The pools GeckoTerminal indexes among `addresses`; pools it never saw are simply absent. */
  async pools(addresses: string[]): Promise<GeckoPool[]> {
    const out: GeckoPool[] = [];
    for (let i = 0; i < addresses.length; i += BATCH) {
      const body = await this.get<{ data?: Resource<{ address: string; name: string; pool_created_at: string; fdv_usd?: string | null }>[] }>(
        `/pools/multi/${addresses.slice(i, i + BATCH).join(',')}`,
      );
      for (const p of body?.data ?? []) {
        out.push({
          address: p.attributes.address.toLowerCase(),
          dex: p.relationships?.dex?.data?.id ?? '',
          createdAt: p.attributes.pool_created_at,
          baseToken: stripNetwork(p.relationships?.base_token?.data?.id),
          quoteToken: stripNetwork(p.relationships?.quote_token?.data?.id),
          name: p.attributes.name,
          fdvUsd: Number(p.attributes.fdv_usd) || 0,
        });
      }
    }
    return out;
  }

  async tokens(addresses: string[]): Promise<GeckoToken[]> {
    const out: GeckoToken[] = [];
    for (let i = 0; i < addresses.length; i += BATCH) {
      const body = await this.get<{
        data?: Resource<{ address: string; name: string; symbol: string; decimals: number; normalized_total_supply: string | number; image_url?: string | null }>[];
      }>(`/tokens/multi/${addresses.slice(i, i + BATCH).join(',')}`);
      for (const t of body?.data ?? []) {
        const a = t.attributes;
        const image = a.image_url && /^https:\/\//.test(a.image_url) && !a.image_url.includes('missing') ? a.image_url : undefined;
        out.push({ address: a.address.toLowerCase(), name: a.name, symbol: a.symbol, decimals: Number(a.decimals), supply: Number(a.normalized_total_supply), imageUrl: image });
      }
    }
    return out;
  }

  /** Hourly USD closes of `token` in `pool` before `beforeSec`, as [unix seconds, close], newest first. */
  async hourlyCloses(pool: string, token: string, beforeSec: number, limit = 1000): Promise<[number, number][]> {
    const body = await this.get<{ data?: { attributes?: { ohlcv_list?: [number, number, number, number, number, number][] } } }>(
      `/pools/${pool}/ohlcv/hour?before_timestamp=${beforeSec}&limit=${limit}&currency=usd&token=${token}`,
    );
    return (body?.data?.attributes?.ohlcv_list ?? []).map(([t, , , , close]) => [t, close]);
  }

  /** The pool GeckoTerminal lists first for a token (its deepest market), or null. */
  async topPool(token: string): Promise<string | null> {
    const body = await this.get<{ data?: Resource<{ address: string }>[] }>(`/tokens/${token}/pools?page=1`);
    return body?.data?.[0]?.attributes.address.toLowerCase() ?? null;
  }

  /** Highest hourly USD price of `token` in `pool` between two unix times (seconds), or null without trades. */
  async peakPrice(pool: string, token: string, fromSec: number, toSec: number): Promise<number | null> {
    const hours = Math.min(1000, Math.ceil((toSec - fromSec) / 3600) + 2);
    const body = await this.get<{ data?: { attributes?: { ohlcv_list?: [number, number, number, number, number, number][] } } }>(
      `/pools/${pool}/ohlcv/hour?before_timestamp=${toSec}&limit=${hours}&currency=usd&token=${token}`,
    );
    const candles = (body?.data?.attributes?.ohlcv_list ?? []).filter(([t]) => t >= fromSec - 3600 && t <= toSec);
    if (!candles.length) return null;
    return candles.reduce((max, [, , high]) => Math.max(max, high), 0);
  }
}

File: backend/wassily/chain/gecko.py
"""GeckoTerminal's public API (no key).

The free tier allows about 30 calls a minute, but a shared cloud IP often gets
less, so the pace adapts: it slows down on every 429 and creeps back up while
calls succeed.
"""

from __future__ import annotations

import asyncio
import time
from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass, field
from typing import Any

import httpx

from wassily.utils.logging import get_logger

log = get_logger("gecko")

BATCH = 30  # the multi endpoints accept up to 30 addresses
MIN_PER_MINUTE = 4.0


class GeckoUnavailable(RuntimeError):
    """Unanswered is not the same as "no data": callers must retry later instead of drawing conclusions."""


@dataclass(frozen=True)
class GeckoPool:
    address: str  # as GeckoTerminal lists it: pair/pool address, or the v4 pool id
    dex: str
    created_at: str
    base_token: str
    quote_token: str
    name: str
    fdv_usd: float  # current fully diluted value of the base token


@dataclass(frozen=True)
class GeckoToken:
    address: str
    name: str
    symbol: str
    decimals: int
    supply: float  # normalized total supply
    image_url: str | None = None


@dataclass
class GeckoStats:
    calls: int = 0
    throttled: int = 0
    failed: int = 0
    per_minute: float = 0.0
    last_error: str | None = field(default=None)


def _strip_network(resource_id: str | None) -> str:
    rid = resource_id or ""
    return rid[rid.find("_") + 1 :].lower()


class GeckoClient:
    def __init__(
        self,
        network: str,
        *,
        api: str = "https://api.geckoterminal.com/api/v2",
        per_minute: float = 28,
        http: httpx.AsyncClient | None = None,
        sleep: Callable[[float], Awaitable[None]] = asyncio.sleep,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        self.network = network
        self.api = api.rstrip("/")
        self._http = http or httpx.AsyncClient(timeout=30.0, headers={"accept": "application/json"})
        self._owns_http = http is None
        self._sleep = sleep
        self._clock = clock
        self._ceiling = per_minute
        self._pace = per_minute
        self._next_at = 0.0
        self.stats = GeckoStats(per_minute=per_minute)

    async def aclose(self) -> None:
        if self._owns_http:
            await self._http.aclose()

    def backlog_seconds(self) -> float:
        """How long a call made now would wait for its turn."""
        return max(0.0, self._next_at - self._clock())

    async def _get(self, path: str) -> Any | None:
        url = f"{self.api}/networks/{self.network}{path}"
        for attempt in range(6):
            at = max(self._clock(), self._next_at)
            self._next_at = at + 60.0 / self._pace
            wait = at - self._clock()
            if wait > 0:
                await self._sleep(wait)
            self.stats.calls += 1
            try:
                res = await self._http.get(url)
                if res.status_code == 429:
                    self.stats.throttled += 1
                    self._pace = max(MIN_PER_MINUTE, self._pace * 0.7)
                    self._next_at = max(self._next_at, self._clock() + 10.0)
                    self.stats.per_minute = round(self._pace, 1)
                    continue
                self._pace = min(self._ceiling, self._pace + 0.2)
                self.stats.per_minute = round(self._pace, 1)
                if res.status_code == 404:
                    return None
                if res.status_code >= 400:
                    raise RuntimeError(f"HTTP {res.status_code}")
                return res.json()
            except (httpx.HTTPError, RuntimeError, ValueError) as err:
                self.stats.failed += 1
                self.stats.last_error = str(err)
                log.warning("geckoterminal %s: %s", path[:60], err)
                self._next_at = max(self._next_at, self._clock() + 5.0 * (attempt + 1))
        raise GeckoUnavailable(f"geckoterminal unavailable for {path[:60]}")

    async def pools(self, addresses: Sequence[str]) -> list[GeckoPool]:
        """The pools GeckoTerminal indexes among ``addresses``; pools it never saw are simply absent."""
        out: list[GeckoPool] = []
        for i in range(0, len(addresses), BATCH):
            body = await self._get(f"/pools/multi/{','.join(addresses[i : i + BATCH])}")
            for p in (body or {}).get("data") or []:
                a = p.get("attributes", {})
                rel = p.get("relationships") or {}
                out.append(
                    GeckoPool(
                        address=str(a.get("address", "")).lower(),
                        dex=((rel.get("dex") or {}).get("data") or {}).get("id", ""),
                        created_at=a.get("pool_created_at", ""),
                        base_token=_strip_network(((rel.get("base_token") or {}).get("data") or {}).get("id")),
                        quote_token=_strip_network(((rel.get("quote_token") or {}).get("data") or {}).get("id")),
                        name=a.get("name", ""),
                        fdv_usd=float(a.get("fdv_usd") or 0),
                    )
                )
        return out

    async def tokens(self, addresses: Sequence[str]) -> list[GeckoToken]:
        out: list[GeckoToken] = []
        for i in range(0, len(addresses), BATCH):
            body = await self._get(f"/tokens/multi/{','.join(addresses[i : i + BATCH])}")
            for t in (body or {}).get("data") or []:
                a = t.get("attributes", {})
                image = a.get("image_url")
                if not (isinstance(image, str) and image.startswith("https://") and "missing" not in image):
                    image = None
                out.append(
                    GeckoToken(
                        address=str(a.get("address", "")).lower(),
                        name=a.get("name", ""),
                        symbol=a.get("symbol", ""),
                        decimals=int(a.get("decimals") or 0),
                        supply=float(a.get("normalized_total_supply") or 0),
                        image_url=image,
                    )
                )
        return out

    async def _ohlcv(self, pool: str, token: str, before_sec: int, limit: int) -> list[list[float]]:
        body = await self._get(
            f"/pools/{pool}/ohlcv/hour?before_timestamp={before_sec}&limit={limit}&currency=usd&token={token}"
        )
        return ((((body or {}).get("data") or {}).get("attributes")) or {}).get("ohlcv_list") or []

    async def hourly_closes(self, pool: str, token: str, before_sec: int, limit: int = 1000) -> list[tuple[int, float]]:
        """Hourly USD closes of ``token`` in ``pool`` before ``before_sec``, as (unix seconds, close), newest first."""
        return [(int(c[0]), float(c[4])) for c in await self._ohlcv(pool, token, before_sec, limit)]

    async def top_pool(self, token: str) -> str | None:
        """The pool GeckoTerminal lists first for a token (its deepest market), or None."""
        body = await self._get(f"/tokens/{token}/pools?page=1")
        data = (body or {}).get("data") or []
        return str(data[0]["attributes"]["address"]).lower() if data else None

    async def peak_price(self, pool: str, token: str, from_sec: int, to_sec: int) -> float | None:
        """Highest hourly USD price of ``token`` in ``pool`` between two unix times, or None without trades."""
        hours = min(1000, -(-(to_sec - from_sec) // 3600) + 2)
        candles = [c for c in await self._ohlcv(pool, token, to_sec, hours) if from_sec - 3600 <= c[0] <= to_sec]
        if not candles:
            return None
        return max(float(c[2]) for c in candles)
